### ISSUE TITLE
Error with using `shape` when parsing expression

### ISSUE THREAD

**Issue by ericpre** [sympy/sympy#21291]
Not sure if this is an issue but the following worked with sympy 1.7.1 and not anymore with 1.8:

```python
import sympy as sp

expr = sp.parse_expr("shape * exp(x)")
```
Give the following error:
```python
ValueError: Error from parse_expr with transformed code: "shape *exp (Symbol ('x' ))"


The above exception was the direct cause of the following exception:

Traceback (most recent call last):

  File "<ipython-input-1-7e220fc8363d>", line 3, in <module>
    expr = sp.parse_expr("shape * exp(x)")

  File "/home/eric/Dev/others/sympy/sympy/parsing/sympy_parser.py", line 1018, in parse_expr
    raise e from ValueError(f"Error from parse_expr with transformed code: {code!r}")

  File "/home/eric/Dev/others/sympy/sympy/parsing/sympy_parser.py", line 1016, in parse_expr
    return eval_expr(code, local_dict, global_dict)

  File "/home/eric/Dev/others/sympy/sympy/parsing/sympy_parser.py", line 910, in eval_expr
    expr = eval(

  File "<string>", line 1, in <module>

TypeError: unsupported operand type(s) for *: 'function' and 'exp'
```

Bisection shows that the failure have been introduced in https://github.com/sympy/sympy/pull/20791. Is it an expected behaviour?

See https://github.com/hyperspy/hyperspy/pull/2701 for background.

**Comment by oscarbenjamin** [comment]
This is a problem with the basic design of `parse_expr`. Essentially any "word" in the input might be the name of a sympy function or otherwise it is considered to be a symbol. SymPy 1.8 has added a new function `sympy.shape` and that now conflicts with using `parse_expr` for any expression where `shape` is expected to be parsed as a symbol. Of course every release of sympy is probably going to add some top-level name so with that design it isn't possible to prevent cross-version breakage in parsing.

The solution is to provide a namespace explicitly:
```python
In [8]: import sympy as sym

In [9]: namespace = {'exp':sym.exp, 'Symbol':sym.Symbol}

In [10]: parse_expr('shape * exp(x)', global_dict=namespace)
Out[10]: 
       x
shape⋅ℯ 
```
With that you can explicitly specify what you want to be considered as a "built-in" function compared to what should be treated as a symbol.

A simple fix would be for sympy to provide some premade namespaces that would be guaranteed not to change from one version to another. There are something like 1000 names in the sympy namespace and most applications won't want most of them.

The fact that `Symbol` needs to be in the namespace to parse a symbol also seems like a bug to me.

**Comment by ericpre** [comment]
Thanks for the explanation, it clarifies what is going on!

**Comment by oscarbenjamin** [comment]
I'm reopening this because I think that there are things to be fixed. I don't think this is a regression in the sense that it should be fixed in a bugfix release. I do think though that there is a problem in the basic design of `parse_expr` here but there is a relatively easy to fix that could be made and added to the documentation.

Basically we should add a limited namespace of common mathematical functions and symbols. I'm fairly sure most uses of `parse_expr` just want to parse simple mathematical expressions like `cos(x) + 1` rather than sympy operations like `solve(x**2 - 1)`. We should add a namespace that just lists these things and is not expected to change from one sympy release to another any time a new function is added to sympy. The docstring for `parse_expr` does mention this but it could be more explicit about why you might want to pass a namespace and there should be an easier way to get a "standard" namespace since the default one is not suitable for most cases.

The namespace could be a dict giving an explicit list of functions like these:
```
abs,arg,
sqrt, cbrt,
exp, log,
sin, cos, tan, 
sec,csc,cot,
sinh,cosh,tanh,
sech,csch,coth,
asin,acos,...
```
Perhaps there could be multiple dicts like:
```
from sympy.parsing.namespaces import elementary_functions, special_functions

expr = parse_expr(string, global_dict=elementary_functions | special_functions)
```
(In Python 3.9 dicts can be combined with `|`)

**Comment by asmeurer** [comment]
This seems like a bigger question of how to organize the names in the default namespace into user presented submodules. Right now, functions have a submodule organization, but it isn't really considered part of the public API if the function is in the main `sympy.__all__`. And it also ties in to how to organize things that aren't in the public namespace, and how to indicate which of those are considered public vs. private API, which has been discussed previously in other issues.

**Comment by oscarbenjamin** [comment]
I think that this is very much separate from how to organise names to be accessible for import in Python. You might be thinking of `parse_expr` as being a way of executing Python code but that's not how most users see it.

The suggestion here is just that there should be a good way to specify what the defined names/functions should be in a DSL for parsing strings. The fact that `parse_expr` uses the `sympy.__all__` namespace by default is just lazy design (and a bad idea).

**Comment by asmeurer** [comment]
The dicts you suggested are basically namespaces, i.e., modules. The special and elementary functions in SymPy are already organized like that in `sympy.functions.elementary` and `sympy.functions.special`. So the question is, do we want the user-level organization of functions to match the actual organization in the module? It's a lot cleaner if it does, and the only downside is it makes the module organization harder to change because it becomes part of the API. But your general idea about de-flattening the namespace implies doing that anyway. 

In other words, I think a simple way to handle this would be to pass a list of submodules to parse_expr, from which it constructs a namespace. There could also be a pre-made list of submodules that constitute the entirety of the top-level sympy namespace.

**Comment by asmeurer** [comment]
This is also related to https://github.com/sympy/sympy/issues/21226, because currently it's not simple to actually disable the transformation that converts undefined names into symbols.

**Comment by oscarbenjamin** [comment]
> In other words, I think a simple way to handle this would be to pass a list of submodules to parse_expr, from which it constructs a namespace.

I think that this is a bad idea and goes precisely against what I am suggesting in this issue. We should recognise that users want to use `parse_expr` in order to parse expressions written in some kind of language and do not care at all about sympy's own organisation of the codebase. The language that they want to parse is usually not Python syntax let alone SymPy code.

I suggested "elementary" and "special" as names but I really didn't mean that to refer to the contents of those parts of the codebase. What I intended was that we could define a documented curated set of functions (and symbols like pi etc) that users might want for parsing "normal" expressions. There could also be a more extended version that includes less common functions like `beta`, `W` etc. These functions wouldn't need to have the same names as the sympy functions/classes (e.g. `LambertW`). Their names should be dictated by what makes sense in parsing applications.

The problem I want to solve by this is that by default `parse_expr` includes the entire sympy namespace but most users of `parse_expr` want a small subset of that. Currently there's no way for them to get that without explicitly constructing their own namespace dict and going through deciding what to include in it (there are 900 names in `sympy.__all__`!). At minimum the docs should show an example of how to make a reasonable dict that covers most "ordinary" functions and symbols. Ideally it would also show an example that has more than most users would want but is still much smaller than the main sympy namespace. That would mean that a user could choose to use either of those or could easily adapt them.

**Comment by asmeurer** [comment]
I agree that organizing the functions in SymPy in this way is a good idea. The docs should also use this organization. 

But why is it a bad idea to also organize the code along these lines? Is it because it's better to organize the code based on implementation details, which may or may not coincide with the way users think of function categories? Or do you want to leave open the option to reorganize the code without breaking API?

**Comment by oscarbenjamin** [comment]
> But why is it a bad idea to also organize the code along these lines?

I mean that it's a bad idea for the behaviour of `parse_expr` to depend on the way that sympy's code is organised. The point of adding a "standard namespace" would be that it is designed in a way that is useful for particular parsing applications and then kept unchanged from release to release regardless of what happens in the sympy codebase.

Separately the sympy codebase could be better organised. We need an import namespace that is not just `from sympy import X` or `from sympy.package.func_module import func`. There should be a separation between how users import public API and how sympy organises its code and just putting everything in `sympy.__all__` doesn't work. I can see two approaches:

1. Make a `sympy.public` (and `sympy.public.x.y` etc.) namespace that provides stable imports. Clearly document that everything should be imported from there including things that are not available as `from sympy import X`.
2. Make a `sympy._src` package and move all the significant code in there. Leave the other modules as the equivalent of `sympy.public` where they only define public top-level functions to be imported.
